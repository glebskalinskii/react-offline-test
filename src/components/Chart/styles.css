.chart-text {
  font-size: 32px;
}
