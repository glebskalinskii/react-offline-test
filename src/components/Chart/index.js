import React from 'react';
import { PieChart, Pie, Sector, Cell } from 'recharts';

import './styles.css';

const Chart = ({ data, activeIndex, setActiveIndex }) => {
  const filteredData = data.filter(energyUnit => energyUnit.perc > 0);

  const renderActiveShape = (props) => {
    const { cx, cy, midAngle, innerRadius, outerRadius, startAngle, endAngle,
      fill, payload, percent, value } = props;

    return (
      <g>
        <text className="chart-text" x={cx} y={cy} dy={8} textAnchor="middle" fill={fill}>{`${payload.perc}%`}</text>
        <Sector
          cx={cx}
          cy={cy}
          innerRadius={innerRadius}
          outerRadius={outerRadius + 15}
          startAngle={startAngle}
          endAngle={endAngle}
          fill={fill}
        />
      </g>
    );
  };

  const mouseEnterHandler = (data, index) => setActiveIndex(index);
  const mouseLeaveHandler = () => setActiveIndex(null);

  return <PieChart width={500} height={500}>
    <Pie 
      dataKey="perc"
      nameKey="fuel"
      data={filteredData} 
      cx={250} 
      cy={250} 
      innerRadius={100}
      outerRadius={200} 
      activeShape={renderActiveShape}
      activeIndex={activeIndex}
      onMouseEnter={mouseEnterHandler}
      onMouseLeave={mouseLeaveHandler}
      stroke="none"
      paddingAngle={2}
    >
      {
        data.map((energyUnit, index) => <Cell key={energyUnit.fuel} fill={energyUnit.colorKey}/>)
      }
  </Pie>
  </PieChart>
}

export default Chart;
