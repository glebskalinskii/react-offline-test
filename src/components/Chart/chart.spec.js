import React from 'react';
import { shallow } from 'enzyme';
import Chart from './index';
import testData from '../../../example_api_response.json';

describe('Chart', () => {
  test('should render and match snapshot', () => {
    const wrapper = shallow(<Chart
        data={testData.data.generationmix}
        activeIndex={null}
        setActiveIndex={() => {}}
      />);

    expect(wrapper).toMatchSnapshot();
  });
});
