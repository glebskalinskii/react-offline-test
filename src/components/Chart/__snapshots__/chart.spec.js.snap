// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart should render and match snapshot 1`] = `
<PieChart
  barCategoryGap="10%"
  barGap={4}
  cx="50%"
  cy="50%"
  endAngle={360}
  height={500}
  innerRadius={0}
  layout="centric"
  margin={
    Object {
      "bottom": 5,
      "left": 5,
      "right": 5,
      "top": 5,
    }
  }
  outerRadius="80%"
  reverseStackOrder={false}
  stackOffset="none"
  startAngle={0}
  width={500}
>
  <Pie
    activeIndex={null}
    activeShape={[Function]}
    animationBegin={400}
    animationDuration={1500}
    animationEasing="ease"
    blendStroke={false}
    cx={250}
    cy={250}
    data={
      Array [
        Object {
          "fuel": "biomass",
          "perc": 4.8,
        },
        Object {
          "fuel": "coal",
          "perc": 2.5,
        },
        Object {
          "fuel": "imports",
          "perc": 8.7,
        },
        Object {
          "fuel": "gas",
          "perc": 46.5,
        },
        Object {
          "fuel": "nuclear",
          "perc": 16.1,
        },
        Object {
          "fuel": "other",
          "perc": 0.3,
        },
        Object {
          "fuel": "hydro",
          "perc": 0.9,
        },
        Object {
          "fuel": "solar",
          "perc": 14.6,
        },
        Object {
          "fuel": "wind",
          "perc": 5.6,
        },
      ]
    }
    dataKey="perc"
    endAngle={360}
    fill="#808080"
    hide={false}
    innerRadius={100}
    isAnimationActive={true}
    labelLine={true}
    legendType="rect"
    minAngle={0}
    nameKey="fuel"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    outerRadius={200}
    paddingAngle={2}
    startAngle={0}
    stroke="none"
  >
    <Cell
      key="biomass"
    />
    <Cell
      key="coal"
    />
    <Cell
      key="imports"
    />
    <Cell
      key="gas"
    />
    <Cell
      key="nuclear"
    />
    <Cell
      key="other"
    />
    <Cell
      key="hydro"
    />
    <Cell
      key="solar"
    />
    <Cell
      key="wind"
    />
  </Pie>
</PieChart>
`;
